core/src/scene/drawRule.cpp

```diff
@@ -157,53 +157,59 @@ bool DrawRuleMergeSet::match(const Feature& _feature, const SceneLayer& _layer,
 
-        bool visible;
-        if (rule.get(StyleParamKey::visible, visible) && !visible) {
-            return false;
-        }
+    bool visible;
+    if (rule.get(StyleParamKey::visible, visible) && !visible) {
+        return false;
+    }
 
-        bool valid = true;
-        for (size_t i = 0; i < StyleParamKeySize; ++i) {
+    bool valid = true;
+    for (size_t i = 0; i < StyleParamKeySize; ++i) {
 
-            if (!rule.active[i]) {
-                rule.params[i].param = nullptr;
-                continue;
-            }
+        if (!rule.active[i]) {
+            rule.params[i].param = nullptr;
+            continue;
+        }
 
-            auto*& param = rule.params[i].param;
+        auto*& param = rule.params[i].param;
 
-            // Evaluate JS functions and Stops
-            if (param->function >= 0) {
+        // Evaluate JS functions and Stops
+        if (param->function >= 0) {
 
-                // Copy param into 'evaluated' and point param to the evaluated StyleParam.
-                m_evaluated[i] = *param;
-                param = &m_evaluated[i];
+            // Copy param into 'evaluated' and point param to the evaluated StyleParam.
+            m_evaluated[i] = *param;
+            param = &m_evaluated[i];
 
-                if (!ctx.evalStyle(param->function, param->key, m_evaluated[i].value)) {
-                    if (StyleParam::isRequired(param->key)) {
-                        valid = false;
-                        break;
-                    } else {
-                        rule.active[i] = false;
-                    }
+            if (!ctx.evalStyle(param->function, param->key, m_evaluated[i].value)) {
+                if (StyleParam::isRequired(param->key)) {
+                    valid = false;
+                    break;
+                } else {
+                    rule.active[i] = false;
                 }
-            } else if (param->stops) {
-                m_evaluated[i] = *param;
-                param = &m_evaluated[i];
-
-                Stops::eval(*param->stops, param->key, ctx.getKeywordZoom(), m_evaluated[i].value);
             }
+        } else if (param->stops) {
+            m_evaluated[i] = *param;
+            param = &m_evaluated[i];
+
+            Stops::eval(*param->stops, param->key, ctx.getKeywordZoom(), m_evaluated[i].value);
         }
+    }
 
-        return valid;
+    return valid;
 }
 
+void DrawRuleMergeSet::mergeRules(const std::vector<const SceneLayer*>& _layers, const std::string& drawGroupName) {
+    for (auto layer: _layers) {
+        mergeRules(*layer);
+    }
+}
 
-void DrawRuleMergeSet::mergeRules(const SceneLayer& _layer) {
+void DrawRuleMergeSet::mergeRules(const SceneLayer& _layer, const std::string& drawGroupName) {
 
     size_t pos, end = m_matchedRules.size();
 
     for (const auto& rule : _layer.rules()) {
+        if (!drawGroupName.empty() && rule.name != drawGroupName) { continue; }
```

Comment:
Checking names in DrawRuleMergeSet is slightly less work for Marker styling, but slightly more work for styling every non-marker feature. Since there are vastly more non-marker features than marker features, we should optimize for that case.

Efficiency aside, I think the code becomes confusing when responsibilities are mixed like this.